### DISCUSSION TITLE
Accessing mutter libraries

### DISCUSSION THREAD

**Discussion by tmayoff** [gtk-rs/gtk4-rs#1679] [Q&A] [answered]
I'm messing around with creating a drop down window using this library. I'd like to be able to move the window a fixed position on screen, from what I can see in the gnome shell extensions for drop-down terminals and tiling extensions they all use a mutter function `move_frame` through the JS bindings. I'd like to be able to do the same through rust. Is there an alternative method to do this already? If not, does anyone have any guidance on how to get access to those mutter functions in rust. I've started to setup the bindgen stuff, but I'm stuck on an error. I'm on nixos, which might just be complicating everything even more.

```
  /nix/store/z1xzvmw2bzzk3x403zab39lq0r2sxbkw-glib-2.78.4-dev/include/glib-2.0/glib/gmacros.h:40:10: fatal error: 'stddef.h' file not found
  thread 'main' panicked at build.rs:11:10:
  Unable to generate bindings: ClangDiagnostic("/nix/store/z1xzvmw2bzzk3x403zab39lq0r2sxbkw-glib-2.78.4-dev/include/glib-2.0/glib/gmacros.h:40:10: fatal error: 'stddef.h' file not found\n")
  note: run with `RUST_BACKTRACE=1` environment variable to display a 
```

```rust
// build.rs
use std::{env, path::PathBuf};

fn main() {
    let mutter = pkg_config::probe_library("libmutter-13").unwrap();

    let bindings = bindgen::Builder::default()
        .header("src/mutter/wrapper.h")
        .clang_args(mutter.include_paths.iter().map(|path| format!("-I{}", path.to_string_lossy())))
        .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
        .generate()
        .expect("Unable to generate bindings");

    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings");
}
```

**Comment by bilelmoussaoui** [comment]
First of all, This has nothing to do with gtk4-rs. Secondly, libmutter is meant to be used from the compositor (Shell extensions are run by GNOME Shell, which is a compositor built with libmutter) and not by the apps themselves. So what are you trying to achieve would be the proper question to ask.

**Reply by tmayoff** [reply]
Apologies, I came here as I was using this library to build the UI and this org does have [this](https://github.com/gtk-rs/gtk-rs-core)  (maybe I should have opened this there). What I'd like to achieve is positioning the window somewhere specific on screen as well as size it specifically. 

Everywhere I see mentions that this needs to be done via the compositor hence libmutter, and backed by looking into the extensions which call libmutter functions. I hoped that I'd still be able to access those functions or achieve the goal without being a shell extension

**Reply by bilelmoussaoui** [reply, marked as answer]
I see, well when people say this has to be done by the compositor, it doesn't mean the app should use the compositor library for that. Things will break pretty badly imho.

That basically means, the application cannot control their position at all. You should ship a GNOME Shell extension alongside the app if you want such behaviour (and it will only work in GNOME Shell obviously) 

For the window size, just set the default-width/default-height of your window.
